src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/CounterAggregator.cs

```diff
@@ -16,21 +24,56 @@ public CounterAggregator(bool isMonotonic)
 
         public override void Update(double value)
         {
-            lock (this)
+            // Get the deltas array.
+            PaddedDouble[] deltas = _deltas;
+
+            // Get the delta best associated with the current thread, preferring to use core ID rather than
+            // thread ID to reduce contention.
+            ref PaddedDouble delta = ref deltas[
+#if NETCOREAPP2_1_OR_GREATER
+                Thread.GetCurrentProcessorId()
+#else
+                Environment.CurrentManagedThreadId
+#endif
+                % deltas.Length];
+
+            // We're not guaranteed uncontented access, so we still need to add the value
+            // to the delta with synchronization. Contention could come from other threads
+            // assigned to the same slot or from Collect zero'ing out the delta.
+            double currentValue;
+            do
             {
-                _delta += value;
+                currentValue = delta.Value;
             }
+            while (Interlocked.CompareExchange(ref delta.Value, currentValue + value, currentValue) != currentValue);
```

Comment:
Perhaps, worth filing an api proposal for `Interlocked.Add(double)` ?

Reply:
What would it do?

Reply:
The same as what integer overloads do? E.g.:
```cs
public float Add(ref float location, float value)
{
    float initialValue, computedValue;
    do
    {
        initialValue = totalValue;
        computedValue = initialValue + value;
    }
    while (initialValue != Interlocked.CompareExchange(ref totalValue, 
        computedValue, initialValue));
    return computedValue;
}
```

Reply:
The integer Add methods end up in intrinsics that result in a single lock add instruction on some platforms, no?

Reply:
Yes, this won't be accelerated (although, e.g. Add on arm64 8.0 is also not accelerated), I was just thinking of hiding this a bit obscured impl under a simple API if users might needed - no idea if they do 🙂

Reply:
> although, e.g. Add on arm64 8.0 is also not accelerated

Yes, I just only think we should be adding such apis if they're accelerated on at least one platform, especially when they're overloads of methods that are accelerated. Otherwise overloads end up having very different perf characteristics.